The ``101-lazy_matrix_mul`` module
============================

Using ``lazy_matrix_mul(m_a, m_b)``
---------------------

Importing function from the module:

    >>> lazy_matrix_mul = __import__('101-lazy_matrix_mul').lazy_matrix_mul

Normal test:

    >>> print(lazy_matrix_mul([[1, 2], [3, 4]], [[1, 2], [3, 4]]))
    [[ 7 10]
     [15 22]]

Passing m_a as non list:

    >>> lazy_matrix_mul("not a list", [[1, 2], [3, 4]])
    Traceback (most recent call last):
        ...
    ValueError: Scalar operands are not allowed, use '*' instead

Passing m_b as non list:

    >>> lazy_matrix_mul([[1, 2], [3, 4]], "not a list")
    Traceback (most recent call last):
        ...
    ValueError: Scalar operands are not allowed, use '*' instead


Passing m_a as non list of lists:

    >>> lazy_matrix_mul([1, 2], [[1, 2], [3, 4]])
    array([ 7, 10])

Passing m_a as non list of lists:

    >>> lazy_matrix_mul([[1, 2], [3, 4]], [1, 2])
    array([ 5, 11])

Passing m_a empty:

    >>> lazy_matrix_mul([[]], [[1, 2], [3, 4]])
    Traceback (most recent call last):
        ...
    ValueError: shapes (1,0) and (2,2) not aligned: 0 (dim 1) != 2 (dim 0)

Passing m_b empty:

    >>> lazy_matrix_mul([[1, 2], [3, 4]], [])
    Traceback (most recent call last):
        ...
    ValueError: shapes (2,2) and (0,) not aligned: 2 (dim 1) != 0 (dim 0)

Passing m_a with non int or float:

    >>> lazy_matrix_mul([["Hi", "bruh"], [3, 4]], [[1, 2], [3, 4]])
    Traceback (most recent call last):
        ...
    TypeError: invalid data type for einsum

Passing m_b with non int or float:

    >>> lazy_matrix_mul([[1, 2], [3, 4]], [["sup", "dude"], [3, 4]])
    Traceback (most recent call last):
        ...
    TypeError: invalid data type for einsum

Passing matrices that can't be multiplied:

    >>> lazy_matrix_mul([[1, 2], [3, 4]], [[1, 2], [3, 4], [5, 6]])
    Traceback (most recent call last):
        ...
    ValueError: shapes (2,2) and (3,2) not aligned: 2 (dim 1) != 3 (dim 0)

Passing m_a with different sizes:

    >>> lazy_matrix_mul([[1, 2], [3, 4, 5]], [[1, 2], [3, 4]])
    Traceback (most recent call last):
        ...
    ValueError: setting an array element with a sequence.

Passing m_b with different sizes:

    >>> lazy_matrix_mul([[1, 2], [3, 4]], [[1, 2], [3, 4, 5]])
    Traceback (most recent call last):
        ...
    ValueError: setting an array element with a sequence.

Passing 1 Arg:

    >>> lazy_matrix_mul([[1, 2], [3, 4]])
    Traceback (most recent call last):
        ...
    TypeError: lazy_matrix_mul() missing 1 required positional argument: 'm_b'

Passing no Arg:

    >>> lazy_matrix_mul()
    Traceback (most recent call last):
        ...
    TypeError: lazy_matrix_mul() missing 2 required positional arguments: 'm_a' and 'm_b'
